plugins/woocommerce/includes/admin/class-wc-admin-marketplace-promotions.php

```diff
@@ -66,31 +49,73 @@ public static function init() {
 			return;
 		}
 
+		if ( ! is_admin() ) {
+			return;
+		}
+
+		self::maybe_update_promotions();
```

Comment:
Or alternatively, via ajax request

Reply:
@ilyasfoo, we can take a look at this one. Before we do though, do you think it'd make sense to create some abstraction layer making it easier to communicate with WCCOM in a safe, async, and cachable way?

Reply:
Thanks, @mcliwanow!

> do you think it'd make sense to create some abstraction layer making it easier to communicate with WCCOM in a safe, async, and cachable way?

We do have an existing generic class ([DataSourcePoller](https://github.com/woocommerce/woocommerce/blob/0a60c3c5ec9b4830ab7d4e3817354f25bec597c3/plugins/woocommerce/src/Admin/RemoteSpecs/DataSourcePoller.php#L1)) which handles API requests, locale, and caching. However, it doesn't implement async queries out of the box.

Feel free to improve the class if you'd like, I'd be glad to help around with reviews and suggestions!

Reply:
Hey, @andfinally. Thanks for looking into it!

> What's the background to this?

The primary goal of changing it to async is to improve the stability `wp-admin` experience. As of now, the request runs synchronously and runs on all `wp-admin` pages. In the event of an unresponsive `woocommerce.com` and cache not available, it can potentially hold up `wp-admin` requests at least until a cache is set. 

> Are we worried for example that if the endpoint requested no longer exists, this will break the rest of the page?

While it's unlikely for `woocommerce.com` to go down, when it happens, it could stall merchants/admin users even if don't intend to use WooCommerce. In a typical configuration, API requests or PHP execution timeouts can take up to 30 seconds. I think it's still worth attempting to improve the experience.

We can simulate the current issue by using beta tester tool with the following steps:

1. Create a fresh Jurassic Ninja site with `woocommerce`
2. Install `woocommerce-beta-tester` (download [woocommerce-beta-tester.zip](https://github.com/user-attachments/files/17094190/woocommerce-beta-tester.zip) or build from `trunk` in core repo)
3. Go to `Tools > WCA Test Helper > Tools`
4. In `Force errors on woocommerce.com requests`, select `Timeout after 6 seconds` and click `Run` ([screenshot](https://github.com/user-attachments/assets/de3ab026-5bb7-41de-8f5f-c383da9916f1))
5. SSH to the site, and run `wp cache flush`
6. Go to `/wp-admin`
7. Observe the request takes at least 6 seconds to load

Reply:
HI @ilyasfoo thanks for the details! We have an [issue](21774-gh-Automattic/woocommerce.com) to fix this in the upcoming release.